test_conformance/extensions/cl_khr_command_buffer/command_buffer_queue_substitution.cpp

```diff
@@ -252,6 +252,146 @@ struct SubstituteQueueTest : public BasicCommandBufferTest
     clEventWrapper user_event;
 };
 
+// Command-queue substitution tests which handles below cases:
+// * Template param is true - Create a command-buffer with an in-order queue,
+//   and enqueue command-buffer to an out-of-order queue.
+// * Template param is false - Create a command-buffer with an out-of-order
+//   queue, and enqueue command-buffer to an in-order queue.
+template <bool is_ooo_test>
+struct QueueOrderTest : public BasicCommandBufferTest
+{
+    using BasicCommandBufferTest::BasicCommandBufferTest;
+
+    QueueOrderTest(cl_device_id device, cl_context context,
+                   cl_command_queue queue)
+        : BasicCommandBufferTest(device, context, queue), ooo_queue(nullptr),
+          ooo_command_buffer(this)
```

Comment:
I don't quite understand `ooo_command_buffer(this)`.  Should this initialize to `nullptr` instead?  Even better, can we just rely on the wrapper's default constructor behavior?

```suggestion
          ooo_command_buffer(nullptr)
```

Reply:
It's a bit convoluted https://github.com/KhronosGroup/OpenCL-CTS/blob/main/test_conformance/extensions/cl_khr_command_buffer/command_buffer_test_base.h#L87-L91

 The `clCommandBufferWrapper` class can't call the `clRetainCommandBufferKHR` and `clRetainCommandBufferKHR` entrypoints without them being queried from the platform associated with the device, so it takes a pointer to the test instance to get that info from. We do this `command_buffer(this)` initalization too in the `BasicCommandBufferTest` class most tests are derived from, so if we have ideas of changing this it would probably be best tracked in an issue for a standalone PR to address.

Reply:
Ah, gotcha.  Yeah, if we're going to change this, we should do so in a different PR, agreed.